src/librustdoc/clean/types.rs

```diff
@@ -860,10 +821,55 @@ impl Item {
                 out.push(&int_s);
             }
             if !out.is_empty() {
-                attrs.push(format!("#[repr({})]", out.join(", ")));
+                return Some(format!("#[repr({})]", out.join(", ")));
             }
         }
-        attrs
+        None
+    }
+
+    pub(crate) fn attributes(
+        &self,
+        tcx: TyCtxt<'_>,
+        cache: &Cache,
+        is_json: bool,
+    ) -> impl Iterator<Item = String> {
+        const ALLOWED_ATTRIBUTES: &[Symbol] =
+            &[sym::export_name, sym::link_section, sym::no_mangle, sym::non_exhaustive];
+
+        self.attrs
+            .other_attrs
+            .iter()
+            .filter_map(move |attr| {
+                if is_json {
+                    match attr {
+                        hir::Attribute::Parsed(AttributeKind::Deprecation { .. }) => {
+                            // rustdoc-json stores this in `Item::deprecation`, so we
+                            // don't want it it `Item::attrs`.
+                            None
+                        }
+                        rustc_hir::Attribute::Parsed(rustc_attr_parsing::AttributeKind::Repr(
+                            ..,
+                        )) => {
+                            // We have separate pretty-printing logic for `#[repr(..)]` attributes.
+                            // For example, there are circumstances where `#[repr(transparent)]`
+                            // is applied but should not be publicly shown in rustdoc
+                            // because it isn't public API.
+                            None
+                        }
+                        _ => Some(rustc_hir_pretty::attribute_to_string(&tcx, attr)),
+                    }
+                } else if attr.has_any_name(ALLOWED_ATTRIBUTES) {
+                    Some(
+                        rustc_hir_pretty::attribute_to_string(&tcx, attr)
+                            .replace("\\\n", "")
+                            .replace('\n', "")
+                            .replace("  ", " "),
+                    )
```

Comment:
I wonder if this makes a measurable improvement?

```suggestion
                    let attr = rustc_hir_pretty::attribute_to_string(&tcx, attr);
                    Some(
                        if attr.contains(['\n', '\\']) || attr.contains("  ") {
                            attr
                                .replace("\\\n", "")
                                .replace('\n', "")
                                .replace("  ", " ")
                        } else {
                            attr
                        }
                    )
```